clusterctl/cmd/validate_cluster.go

```diff
@@ -34,9 +50,113 @@ var validateClusterCmd = &cobra.Command{
 }
 
 func init() {
+	validateClusterCmd.Flags().StringVarP(
+		&vco.ClusterName, "cluster-name", "", "",
+		"The name of the cluster to be validated. This can be omitted if there is only one cluster being managed in the given namespace.")
+	validateClusterCmd.Flags().StringVarP(
+		&vco.Namespace, "namespace", "n", v1.NamespaceDefault,
+		"If present, validate the cluster in the specified namespace.")
+	validateClusterCmd.Flags().StringVarP(
+		&vco.Kubeconfig, "kubeconfig", "", "",
+		"The file path of the kubeconfig file for Cluster API management stack. If not specified, $KUBECONFIG environment variable or ${HOME}/.kube/config is used.")
 	validateCmd.AddCommand(validateClusterCmd)
 }
 
 func RunValidateCluster() error {
-	return errors.NotImplementedError
+	kubeconfig := vco.Kubeconfig
+	if kubeconfig == "" {
+		kubeconfig = os.Getenv("KUBECONFIG")
+		if kubeconfig == "" {
+			kubeconfig = util.GetDefaultKubeConfigPath()
+		}
+	}
+	client, err := util.NewClientSet(kubeconfig)
+	if err != nil {
+		return fmt.Errorf("Failed to create client for talking to the apiserver: %v", err)
+	}
+
+	cluster, err := getClusterObject(client, vco.ClusterName, vco.Namespace)
+	if err != nil {
+		return err
+	}
+
+	if err = validateClusterAPIObjects(client, cluster); err != nil {
+		return err
+	}
+
+	// TODO(wangzhen127): Also validate the cluster in addition to the cluster API objects.
+
+	glog.Infof("Pass. The cluster %q in namespace %q is validated.", cluster.Name, vco.Namespace)
+	return nil
+}
+
+func validateClusterAPIObjects(client *clientset.Clientset, cluster *v1alpha1.Cluster) error {
```

Comment:
Not going to join the review since you seem to have enough reviewers. (Do let me know if you want me to formally review).

My 2cents: Please break the actual validation leg work into a different package. cmd.go files should only contain the limited stuff to do with how to input via command line. Ie. flags etc

Reply:
+100

Reply:
Now put into clusterctl/validation.

Should we put them in pkg/validation or staying in clusterctl/ is fine. This depends on whether other code wants to reuse this.